File: pkg/platform/recurring_payment_service.go
package platform

import "context"

// type RecurringPaymentType uint8
type RecurringPaymentChargingMethod uint8

// const (
// 	RecurringPaymentYearly RecurringPaymentType = iota
// 	RecurringPaymentMonthly
// 	RecurringPaymentFixedInterval
// )

const (
	RecurringPaymentChargeUpfront RecurringPaymentChargingMethod = iota
	RecurringPaymentChargeNonUpfront
)

type RecurringPaymentService interface {
	// should return history when subscribing to upfront recurring payments
	Subscribe(ctx context.Context, account *PaymentAccount, billingID, callbackData string) (*RecurringPayment, *TransactionHistory, error)
	// should return history when subscribing to non-upfront recurring payments
	Unsubscribe(ctx context.Context, payment *RecurringPayment) (*TransactionHistory, error)

	Bill(ctx context.Context, payment *RecurringPayment) (*TransactionHistory, error)
}

File: fastify/platform/recurring_payment_service.ts
import type { PaymentAccount } from './payment_account'
import type { RecurringPayment } from './recurring_payment'
import type { TransactionHistory } from './transaction_history'

export enum RecurringPaymentChargingMethod {
	Upfront,
	NonUpfront
}

export interface RecurringPaymentService {
	subscribe(signal: AbortSignal, account: PaymentAccount, billingID: string, callbackData: string): Promise<{ payment: RecurringPayment, history: TransactionHistory }>
	unsubscribe(signal: AbortSignal, payment: RecurringPayment): Promise<TransactionHistory>
	bill(signal: AbortSignal, payment: RecurringPayment): Promise<TransactionHistory>
}